pyccel/ast/headers.py

```diff
@@ -264,11 +265,12 @@ def create_definition(self, templates = ()):
         # TODO factorize what can be factorized
         from itertools import product
 
-        name = str(self.name)
+        name = self.name
 
         body      = []
         cls_name  = None
         is_static = self.is_static
+        used_names = set(name)
```

Comment:
The placement of this line means that we generate the signature as follows:
```
function high_int_int_1(function1, function2, a) result(Out_0009)

  implicit none

  integer(C_INT64_T) :: Out_0009
  integer(C_INT64_T), value :: a
  integer(C_INT64_T) :: x
  integer(C_INT64_T) :: y

  interface
    function function1(in__0000) result(out__0000)
      use, intrinsic :: ISO_C_BINDING
      integer(C_INT64_T) :: out__0000
      integer(C_INT64_T), value :: in__0000
    end function function1

    function function2(in__0001) result(out__0001)
      use, intrinsic :: ISO_C_BINDING
      integer(C_INT64_T) :: out__0001
      integer(C_INT64_T), value :: in__0001
    end function function2
  end interface
```
whereas previously we had:
```
function high_int_int_1(function1, function2, a) result(Out_0009)

  implicit none

  integer(C_INT64_T) :: Out_0009
  integer(C_INT64_T), value :: a
  integer(C_INT64_T) :: x
  integer(C_INT64_T) :: y

  interface
    function function1(in_0) result(out_0)
      use, intrinsic :: ISO_C_BINDING
      integer(C_INT64_T) :: out_0
      integer(C_INT64_T), value :: in_0
    end function function1

    function function2(in_0) result(out_0)
      use, intrinsic :: ISO_C_BINDING
      integer(C_INT64_T) :: out_0
      integer(C_INT64_T), value :: in_0
    end function function2
  end interface
```
(Note the arguments, previously in_0 and in_0, but now in_0 and in_1). I think both make sense, but I was just wondering whether this is a conscious choice?

Reply:
my purpose was just to avoid a name clash between args and func_name

Reply:
In that case should `d.name` be added to this set? (I.e. `function1`, `function2`, currently it just contains `high_int_int_1` I think)

Reply:
by ```d.name``` you mean each function name?

Reply:
Yes, I meand [this `d`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/headers.py#L340)
As I said currently this set contains `high_int_int_1` but it would make sense for it to contain at least `function1`, `function2`

Reply:
but I think we only have the signature of function at this stage? we don't have arguments names that why  we create tmp name at 
```                   
arg_name = 'arg_{0}'.format(str(i))
```

Reply:
Ah you are right :+1: